src/auth.rs

```diff
@@ -153,3 +159,96 @@ pub(crate) async fn installed_flow_authenticator<S: AsRef<[u8]>, P: Into<PathBuf
+    scopes: Vec<String>,
+}
+
+impl ApplicationDefaultCredentialsAuthenticator {
+    pub(crate) async fn from_scopes(scopes: &[&str]) -> Result<Arc<dyn Authenticator>, BQError> {
+        let opts = ApplicationDefaultCredentialsFlowOpts::default();
+        let auth = match YupApplicationDefaultCredentialsAuthenticator::builder(opts).await {
+            ApplicationDefaultCredentialsTypes::InstanceMetadata(auth) => auth.build().await,
+            ApplicationDefaultCredentialsTypes::ServiceAccount(auth) => auth.build().await,
+        };
+
+        match auth {
+            Err(err) => Err(BQError::InvalidApplicationDefaultCredentialsAuthenticator(err)),
+            Ok(auth) => Ok(Arc::new(ApplicationDefaultCredentialsAuthenticator {
+                auth: Some(auth),
+                scopes: scopes.iter().map(|scope| scope.to_string()).collect(),
+            })),
+        }
+    }
+}
+
+#[async_trait]
+impl Authenticator for ApplicationDefaultCredentialsAuthenticator {
+    async fn access_token(&self) -> Result<String, BQError> {
+        Ok(self
+            .auth
+            .clone()
+            .unwrap()
+            .token(self.scopes.as_ref())
+            .await?
+            .token()
+            .ok_or(BQError::NoToken)?
+            .to_string())
+    }
+}
+
+pub(crate) async fn application_default_credentials_authenticator(
+    scopes: &[&str],
+) -> Result<Arc<dyn Authenticator>, BQError> {
+    ApplicationDefaultCredentialsAuthenticator::from_scopes(scopes).await
+}
+
+#[derive(Clone)]
+pub struct AuthorizedUserAuthenticator {
+    auth: Option<YupAuthenticator<HttpsConnector<HttpConnector>>>,
+    scopes: Vec<String>,
+}
+
+impl AuthorizedUserAuthenticator {
+    pub(crate) async fn from_authorized_user_secret(
+        authorized_user_secret: AuthorizedUserSecret,
+        scopes: &[&str],
+    ) -> Result<Arc<dyn Authenticator>, BQError> {
+        let auth = YupAuthorizedUserAuthenticator::builder(authorized_user_secret)
+            .build()
+            .await;
+
+        match auth {
+            Err(err) => Err(BQError::InvalidAuthorizedUserAuthenticator(err)),
+            Ok(auth) => Ok(Arc::new(AuthorizedUserAuthenticator {
+                auth: Some(auth),
+                scopes: scopes.iter().map(|scope| scope.to_string()).collect(),
+            })),
+        }
+    }
+}
+
+#[async_trait]
+impl Authenticator for AuthorizedUserAuthenticator {
+    async fn access_token(&self) -> Result<String, BQError> {
+        Ok(self
+            .auth
+            .clone()
+            .unwrap()
+            .token(self.scopes.as_ref())
+            .await?
+            .token()
+            .ok_or(BQError::NoToken)?
+            .to_string())
+    }
+}
+
+pub(crate) async fn authorized_user_authenticator<S: AsRef<Path>>(
+    secret: S,
+    scopes: &[&str],
+) -> Result<Arc<dyn Authenticator>, BQError> {
+    let authorized_user_secret = yup_oauth2::read_authorized_user_secret(secret).await.unwrap();
```

Comment:
Returning an error in this case will be better.

Reply:
Oh, Thank you. I missed it.

Reply:
I fixed it in this commit.

https://github.com/lquerel/gcp-bigquery-client/pull/46/commits/db1d897fb461b3dfb239520e75fa9f7cef37e754